src/main/java/org/jabref/model/openoffice/rangesort/RangeSortVisual.java

```diff
@@ -0,0 +1,154 @@
+package org.jabref.model.openoffice.rangesort;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.jabref.model.openoffice.uno.NoDocumentException;
+import org.jabref.model.openoffice.uno.UnoScreenRefresh;
+
+import com.sun.star.awt.Point;
+import com.sun.star.lang.WrappedTargetException;
+import com.sun.star.text.XTextDocument;
+import com.sun.star.text.XTextRange;
+import com.sun.star.text.XTextViewCursor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Sort XTextRange values visually (top-down,left-to-right).
+ *
+ * Requires functional XTextViewCursor.
+ *
+ * Problem: for multicolumn layout and when viewing pages side-by-side in LO, the
+ *          (top-down,left-to-right) order interpreted as-on-the-screen: an XTextRange at the top of
+ *          the second column or second page is sorted before an XTextRange at the bottom of the
+ *          first column of the first page.
+ */
+public class RangeSortVisual {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RangeSortVisual.class);
+
+    /**
+     * Sort the input {@code inputs} visually.
+     *
+     * Requires a functional {@code XTextViewCursor}.
+     *
+     * @return The input, sorted by the elements XTextRange and getIndexInPosition.
+     */
+    public static <T> List<RangeSortable<T>> visualSort(List<RangeSortable<T>> inputs,
+                                                        XTextDocument doc,
+                                                        FunctionalTextViewCursor fcursor)
+        throws
+        WrappedTargetException,
+        NoDocumentException {
+
+        final int inputSize = inputs.size();
+
+        if (UnoScreenRefresh.hasControllersLocked(doc)) {
+            final String msg = "visualSort: with ControllersLocked, viewCursor.gotoRange is probably useless";
+            LOGGER.warn(msg);
+            throw new IllegalStateException(msg);
+        }
+
+        XTextViewCursor viewCursor = fcursor.getViewCursor();
+
+        // find coordinates
+        List<Point> positions = new ArrayList<>(inputSize);
+        for (RangeSortable<T> v : inputs) {
+            positions.add(findPositionOfTextRange(v.getRange(), viewCursor));
+        }
+        fcursor.restore(doc);
+
+        if (positions.size() != inputSize) {
+            throw new IllegalStateException("visualSort: positions.size() != inputSize");
+        }
+
+        // order by position
+        ArrayList<ComparableMark<RangeSortable<T>>> set = new ArrayList<>(inputSize);
```

Comment:
nitpick: don't call a `List` `set`, even if it is unlikely someone treats it as a `Set` I'd recommend that you avoid it

Reply:
OK, renamed to `comparableMarks`